src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -40,66 +47,93 @@
 
 public class ChannelFactory implements AutoCloseable {
   private static final Logger logger = LoggerFactory.getLogger(ChannelFactory.class);
+  private static final String PUB_KEY_INFRA = "PKIX";
 
+  private long connRecycleTime;
   private final int maxFrameSize;
   private final int keepaliveTime;
   private final int keepaliveTimeout;
   private final int idleTimeout;
-  private final ConcurrentHashMap<String, ManagedChannel> connPool = new ConcurrentHashMap<>();
   private final CertContext certContext;
+  private final CertWatcher certWatcher;
+
+  @VisibleForTesting
+  public final ConcurrentHashMap<String, ManagedChannel> connPool = new ConcurrentHashMap<>();
+
   private final AtomicReference<SslContextBuilder> sslContextBuilder = new AtomicReference<>();
-  private static final String PUB_KEY_INFRA = "PKIX";
 
-  private abstract static class CertContext {
-    protected abstract boolean isModified();
+  private final ScheduledExecutorService recycler = Executors.newSingleThreadScheduledExecutor();
+
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  @VisibleForTesting
+  public static class CertWatcher implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(CertWatcher.class);
+    private final List<File> targets;
+    private final List<Long> lastReload = new ArrayList<>();
+    private final ScheduledExecutorService executorService =
+        Executors.newSingleThreadScheduledExecutor();
+    private final Runnable onChange;
+
+    public CertWatcher(long pollInterval, List<File> targets, Runnable onChange) {
+      this.targets = targets;
+      this.onChange = onChange;
+
+      for (File ignored : targets) {
+        lastReload.add(0L);
+      }
+
+      executorService.scheduleAtFixedRate(
+          this::tryReload, pollInterval, pollInterval, TimeUnit.SECONDS);
+    }
 
-    protected abstract SslContextBuilder createSslContextBuilder();
+    private void tryReload() {
+      if (needReload()) {
+        onChange.run();
+      }
+    }
 
-    public SslContextBuilder reload() {
-      if (isModified()) {
-        logger.info("reload ssl context");
-        return createSslContextBuilder();
+    private boolean needReload() {
+      boolean needReload = false;
+      // Check all the modification of the `targets`.
+      // If one of them changed, means to need reload.
+      for (int i = 0; i < targets.size(); i++) {
+        long lastModified = targets.get(i).lastModified();
```

Comment:
`lastModified` might raise error due to permission. Maybe we need to catch the whole `needReload` body to give a meaningful error log?

Reply:
Done.